cpp/bench/ann/src/common/cuda_huge_page_resource.hpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+// #include <raft/core/logger.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/detail/error.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <sys/mman.h>
+
+#include <cstddef>
+
+namespace rmm::mr {
+/**
+ * @brief `device_memory_resource` derived class that uses cudaMallocHost/Free for
+ * allocation/deallocation.
+ */
+class cuda_huge_page_resource final : public device_memory_resource {
```

Comment:
Updated the documentation.
> Is there a reason why it is not derived from rmm::mr::host_memory_resource?

The goal is to use this allocator in a [device_container_policy](https://github.com/rapidsai/raft/blob/branch-23.12/cpp/include/raft/core/device_container_policy.hpp#L177), and that requires a `device_memory_resource`.
What we have here is not ideal, more like a quick solution to avoid experimenting / benchmarking with such settings. A related issue https://github.com/rapidsai/raft/issues/1819.